File: Baekjoon/Python/데스 나이트.py
from collections import deque
import sys
input = sys.stdin.readline

N = int(input().rstrip())
r1, c1, r2, c2 = map(int, input().split())

dx = [-1,1,-2,2,-1,1]
dy = [-2,-2,0,0,2,2]

graph = [[-1]*N for _ in range(N)]

def BFS(r1, c1):
    q = deque()
    q.append((c1,r1))
    graph[c1][r1] = 0
    while q:
        y, x = q.popleft()
        for i in range(6):
            rx = x + dx[i]
            ry = y + dy[i]
            if 0<=rx<N and 0<=ry<N and graph[ry][rx] == -1:
                graph[ry][rx] = graph[y][x] + 1
                q.append((ry,rx))

BFS(r1,c1)

print(graph[r2][c2])
    

File: Baekjoon/Java/데스 나이트.java
import java.io.*;
import java.util.*;

public class Main {
    static int ans;
    static BufferedReader br;
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        int r1 = Integer.parseInt(st.nextToken());
        int c1 = Integer.parseInt(st.nextToken());
        int r2 = Integer.parseInt(st.nextToken());
        int c2 = Integer.parseInt(st.nextToken());
        int[][] graph = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                graph[i][j] = -1;
            }
        }
        System.out.println(BFS(r1,c1,n,r2,c2,graph));
    }

    public static int BFS(int r1, int c1, int n, int r2, int c2, int[][] graph) {
        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{c1, r1});
        int[] dx = {-1, 1, -2, 2, -1, 1};
        int[] dy = {-2, -2, 0, 0, 2, 2};
        graph[c1][r1] = 0;
        while (!q.isEmpty()) {
            int[] cur = q.pollFirst();
            int x = cur[1];
            int y = cur[0];
            for (int i = 0; i < 6; i++) {
                int rx = x + dx[i];
                int ry = y + dy[i];
                if (0 <= rx && rx < n && 0 <= ry && ry < n && graph[ry][rx] == -1) {
                    graph[ry][rx] = graph[y][x] + 1;
                    q.offer(new int[]{ry, rx});
                }
            }
        }
        return graph[r2][c2];
    }
}
